rfcs/20190829-tfx-container-component-execution.md

```diff
@@ -0,0 +1,555 @@
+
+Status         | Proposed
+:------------- | :-------
+**Author(s)**  | Ajay Gopinathan (<a href="mailto:ajaygopinathan@google.com">ajaygopinathan@google.com</a>), Hongye Sun (<a href="mailto:hongyes@google.com">hongyes@google.com</a>), Makoto Uchida (<a href="mailto:muchida@google.com">muchida@google.com</a>), Ruoyu Liu (<a href="mailto:ruoyu@google.com">ruoyu@google.com</a>)
+**Sponsor(s)** | Konstantinos Katsiapis (katsiapis@google.com), Pavel Dournov (dournov@google.com), Ramesh Chandra (rameshuc@google.com)
+**Updated**    | 2019-08-29
+
+## Objective
+
+This RFC proposes an orchestrator agnostic way to reliably execute a user’s
+container in the TFX pipeline. The proposal can support:
+
+*   Running an arbitrary container in either local docker environment or remote
+    k8s cluster.
+*   Passing data into the container
+*   Passing output data from the container
+*   Capturing logs from the container
+*   Handling errors and retries
+*   Cancelling the container execution if the pipeline is terminated
+
+## Motivation
+
+Currently, in a TFX pipeline, there is no way to execute a generic container as
+one of its steps. Without this feature, users cannot bring their own containers
+into the pipeline. This blocks following use cases:
+
+*   User already had a docker image and want to run the image as one of the
+    steps in a TFX pipeline.
+*   User wants to use non-python code like R as one of the steps in a TFX
+    pipeline.
+*   User wants to have an isolated python environment for their component code.
+
+This RFC is a follow-up design for
+[Container-based Component RFC](https://github.com/tensorflow/community/pull/146).
+This design defines how to execute the container spec as part of a TFX pipeline.
+The execution may occurs in local docker or in a remote kubernetes cluster.
+
+### Existing solutions
+
+#### Kubeflow Pipeline (KFP) ContainerOp
+
+Today, KFP’s ContainerOp leverages
+[Argo container template API](https://github.com/argoproj/argo/blob/master/pkg/apis/workflow/v1alpha1/workflow_types.go)
+to launch user’s container in a k8s pod. Argo as the orchestrator controls when
+to launch the POD and it uses a sidecar container to report output files back
+and wait for user’s container to complete. We are not proposing to use Argo API
+because of the following reasons:
+
+*   Argo’s API is orchestrator-specific and cannot be ported to Airflow or local
+    runners.
+*   Argo’s API doesn’t provide an extensible way to run custom code before and
+    after POD API which is critical to support metadata tracking and caching
+    features.
+*   Argo doesn’t provide an easy way to recover from user’s transient errors,
+    which is critical in production workload.
+
+#### Airflow k8s pod operator
+
+Airflow supports to launch a k8s pod by an
+[operator](https://github.com/apache/airflow/blob/master/airflow/contrib/operators/kubernetes_pod_operator.py).
+This approach is closer to what we are proposing in the document. However, we
```

Comment:
Compared Existing Solutions is there any attempt to make a full replacement of one or more of those solutions (ie creating a better system versus creating a different system suited to TFX's needs)?  This would not be a requirement but a nice to have.